docs/doc/90-contributing/09-how-to-write-aggregate-functions.md

```diff
@@ -2,135 +2,126 @@
 ---
 
+An aggregate function, or aggregation function, is a function that combines values from multiple rows to form a single summary value. Common aggregate functions include sum, count, avg, and others.
+
 ## How to write aggregate functions
 
 Databend allows us to write custom aggregate functions through rust code.
+
 It's not an easy way because you need to be a rustacean first. Databend has a plan to support writing UDAFs in other languages(like js, web assembly) in the future.
 
 In this section we will talk about how to write aggregate functions in Databend.
 
+## Function Registration
 
-## AggregateFunction trait introduction
+The register function for aggregate functions takes a lowercase function name and an `AggregateFunctionDescription` object as parameters. Each registered function is stored in  `case_insensitive_desc` (HashMap data structure).
 
-All aggregate functions implement `AggregateFunction` trait, and we register them into a global static `FunctionFactory`, the factory is just an index map and the key is the name of the aggregate function.
+The `case_insensitive_combinator_desc` is used to store some combined functions, such as `count_if` and `sum_if` that are combined with the `_if` suffix.
 
-:::tip
-Function name in Databend is case-insensitive.
-:::
-
-``` rust
-pub trait AggregateFunction: fmt::Display + Sync + Send {
-    ...
+```rust
+pub struct AggregateFunctionFactory {
+    case_insensitive_desc: HashMap<String, AggregateFunctionDescription>,
+    case_insensitive_combinator_desc: Vec<(String, CombinatorDescription)>,
 }
+impl AggregateFunctionFactory {
+  ...
+pub fn register(&mut self, name: &str, desc: AggregateFunctionDescription) {
+        let case_insensitive_desc = &mut self.case_insensitive_desc;
+        case_insensitive_desc.insert(name.to_lowercase(), desc);
+    }
+  ...
+}
+```
 
-### Understand the functions
-
-- The function `name` indicates the name of this function, such as `sum`, `min`.
-- The function `return_type` indicates the return type of the function, it may vary with different arguments, such as `sum(int8)` -> `int64`, `sum(uint8)` -> `uint64`, `sum(float64)` -> `float64`.
-- The function `nullable` indicates whether the `return_type` is nullable or not.
-
-Before we start to introduce the function `init_state`, let's ask a question first:
-
->  what's aggregate function state?
-
-It indicates the temporary results of an aggregate function. Because an aggregate function accumulates data in columns block by block and there will be some intermediate results after the aggregation. Therefore, the state must be mergeable, serializable.
-
-For example, in the `avg` aggregate function, we can represent the state like:
+Each registered function must implement both the `AggregateFunction` trait and the `AggregateFunctionFeatures` trait, where `AggregateFunctionFeatures` is more similar to `FunctionProperty` in `Scalar` in that they both store some properties of functions.
```

Comment:
... both the xxx and xxx traits, and the `AggregateFunctionFeatures` is more similar to `FunctionProperty` in `Scalar` because they both store some properties of functions.